presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -139,26 +212,32 @@ public static ZOrderCurveSearcher createZOrderCurveSearcher(ZOrder zOrder, List<
      * @param startAddress the starting address of the current Z-curve level
      * @param totalCurveBits the amount of bits in the current Z-curve level
      */
-    private void findAddressesRecursively(int level, long startAddress, int totalCurveBits, List<ZAddressRange<Long>> addressRanges, List<ZValueRange> valueRanges, RangeType rangeType)
+    private void findAddressesRecursively(
+            int level, long startAddress, int totalCurveBits,
+            List<ZAddressRange<Long>> addressRanges, List<ZValueRange> valueRanges, RangeType rangeType, int[] zRangeIndexes)
     {
         long endAddress = startAddress + ((long) 1 << totalCurveBits) - 1;
 
         List<Integer> startValues = zOrder.decode(startAddress);
         List<Integer> endValues = zOrder.decode(endAddress);
 
-        if (inRange(startValues, endValues, valueRanges)) {
+        if (inRange(startValues, endValues, valueRanges, zRangeIndexes)) {
             addToAddressList(startAddress, endAddress, addressRanges, rangeType);
             return;
         }
 
-        if (outOfRange(startValues, endValues, valueRanges)) {
+        if (outOfRange(startValues, endValues, valueRanges, zRangeIndexes)) {
             return;
         }
 
+        if (level < 0 || level >= maxBitLength) {
+            throw new RuntimeException(format("l: %d, d: %d; tbl: %d, r: %d", level, maxBitLength, totalBitLength, valueRanges.size()));
```

Comment:
We should probably check this first in `findAddressesRecursively` given both `level` and `maxBitLength` are fixed after entering the function?

Also, we might want to do `checkArgument`, also let's make the error message more informative (something like `Level out of range. level: %s, expected range: 0 - %s`).

Reply:
Oh, right! That was meant for debugging, but it is a good idea to put in a `checkArgument`!

Reply:
If we are sure it would not happen we could also just remove it :) Up to you.